### DISCUSSION TITLE
Reduce to a palette of n colors of which m are already chosen

### DISCUSSION THREAD

**Discussion by hackerb9** [ImageMagick/ImageMagick#4438] [Help]
Hi! I've been using ImageMagick to convert images to "sixel" to try out on a vintage Digital Equipment VT340 terminal and it (mostly) works. It supports 16 indexed colors with a depth of 4 (4096 possible colors) and it is easy to get ImageMagick to create an image that meets that criteria:

```bash
convert logo: -depth 4 +dither -colors 16 sixel:-
```

Unfortunately, that is not sufficient. I am having the darndest time figuring out how to tell ImageMagick that, although the palette size is 16, three of the colors have already been chosen. 

The problem is that, while all of the colors can be reassigned, three specific ones are used for text 

Index | Typical<br/>Color | Usage
--------|---------|----------
7        |  gray50 | foreground text
15      |  gray75 | bright foreground
0        | black     | background

As you can see in the following screenshot, those colors are getting changed so my text is gold on teal instead of gray on black:

![image](https://user-images.githubusercontent.com/10174223/140246689-71e23465-baff-4f6a-add6-48a77c8355cb.png)

I have read [the ImageMagick docs](https://legacy.imagemagick.org/Usage/quantize/#colors) on quantization and there is an old, untested suggestion that maybe one could add large swaths of preferred colors, but this seems rather gross. I'm hoping that by now there is, if not a better way, at least someone who has tried it and can say how and if that idea works.

____

Note: Eventually, I will need the sixel output to define the colors in the palette in a certain order. I am not asking about that just yet as I imagine it is a very different question.

**Comment by snibgo** [comment]
What version of IM, on what platform (bash, Windows, RSTS/E, VMS)?

If you have IM v7, I suggest "magick", not "convert".

If you want to reduce an image to 4096 colours, where they are evenly spread in the obvious way, 16 values per RGB channel, then:
```
magick in.png -posterize 16 out.png
```
"-colors 16" will reduce that to 16 colours, but IM will choose the "best" colours from the selection of billions. So the next step is to remap the image to the 4096 colours found by "-posterize". This means that each of the 16 colours will be replaced by the nearest of the 4096. Windows CMD syntax:
```
magick ^
  in.png ^
  -posterize 16 ^
  +write mpr:4096 ^
  -colors 16 ^
  -remap mpr:4096 ^
  out.png
```
For bash syntax, change line-continuations from ^ to backslash \\. For RSTS/E or VMS, I dunno.

This will created dithered output. If you don't want that, insert "+dither" after reading the input.

Of course, this doesn't guarantee that any of your three specific colours will be used. Perhaps they were not used in your input image, so won't be on the list of 4096.

If you wanted, you could "edit" the list of 16, replacing the three least-used by the three spefic colours, and remap again.

**Reply by hackerb9** [reply]
Thanks. The heart of my question is: how can I perform both `-remap` where I can specify three colors and also `-colors 16` so 13 other colors will be selected automatically. Currently, I'm trying variations on this:

    convert logo:  -size 640x480 \
        xc:black xc:gray50 xc:gray75 +append \
        +dither -depth 4 -colors 16 \
        -crop 640x480+0+0 sixel:-

I happen to be trying with ImageMagick 6 on Unix, but I'll use whatever works.

**Comment by snibgo** [comment]
You can select the 13 best colours, and add your three to that list, and remap to those 16. (For some reason, this gives 17 colours. So use "-colors 12" instead.)

```
magick ^
  logo: ^
  ( +clone ^
    -colors 12 ^
    xc:black xc:gray50 xc:gray75 +append ^
    -unique-colors ^
    -write mpr:MYMAP ^
    +delete ^
  )^
  -remap mpr:MYMAP ^
  out.png
```

**Reply by hackerb9** [reply]
[The 17c thing seems to be a bug in ImageMagick's PNG writer. If you use png8:out.png you should get 16 colors instead of 17.]

**Comment by snibgo** [comment]
"png8:out.png" doesn't help. I still get 17. But that was a mistake by me. We need to insert "-unique-colours" before appending the the single-pixel images, otherwise IM creates padding.
```
magick ^
  logo: ^
  ( +clone ^
    -colors 13 ^
    -unique-colors ^
    xc:black xc:gray50 xc:gray75 +append ^
    -unique-colors ^
    -write mpr:MYMAP ^
    +delete ^
  )^
  -remap mpr:MYMAP ^
  out.png
```

The resulting 16 colours are:
```
f:\web\im>%IMG7%magick out.png -unique-colors txt:
# ImageMagick pixel enumeration: 16,1,255,srgb
0,0: (0,0,0)  #000000  black
1,0: (9,11,12)  #090B0C  srgb(9,11,12)
2,0: (34,38,101)  #222665  srgb(34,38,101)
3,0: (87,78,73)  #574E49  srgb(87,78,73)
4,0: (127,127,127)  #7F7F7F  grey50
5,0: (252,6,6)  #FC0606  srgb(252,6,6)
6,0: (209,98,69)  #D16245  srgb(209,98,69)
7,0: (121,130,118)  #798276  srgb(121,130,118)
8,0: (244,229,50)  #F4E532  srgb(244,229,50)
9,0: (230,158,92)  #E69E5C  srgb(230,158,92)
10,0: (34,61,145)  #223D91  srgb(34,61,145)
11,0: (63,85,135)  #3F5587  srgb(63,85,135)
12,0: (155,156,159)  #9B9C9F  srgb(155,156,159)
13,0: (191,191,191)  #BFBFBF  grey75
14,0: (215,189,184)  #D7BDB8  srgb(215,189,184)
15,0: (255,255,255)  #FFFFFF  white
```

**Reply by hackerb9** [reply]
Yes, that does somewhat work, but ImageMagick is ignoring the three prechosen colors during quantization and could easily choose a very similar color. So, for example, if I have an image that is mostly black, two of the palette choices are (almost) black. 

```bash
convert logo: -negate \
  \( +clone -colors 13 xc:black xc:gray50 xc:gray75 +append -write mpr:MYMAP +delete \) \
  -remap mpr:MYMAP sixel:- \
  | awk -F';' -v RS='#' \
      'BEGIN {printf("Index:  Red  Green  Blue\n")}
       /^[0-9]+;/ {printf("%5d: %3d%%, %3d%%, %3d%%\n", (NR-1)%16, $3, $4, $5)}' 
```
Output:
```
Index:  Red  Green  Blue
    1:   0%,   0%,   0%
    2:   2%,   3%,   3%
    3:  16%,  26%,  28%
    4:  39%,  39%,  38%
    5:  50%,  50%,  50%
    6:  75%,  67%,  47%
    7:  87%,  76%,  43%
    8:  10%,  38%,  64%
    9:   4%,  10%,  80%
   10:  53%,  49%,  54%
   11:  18%,  62%,  73%
   12:   1%,  98%,  98%
   13:  66%,  69%,  71%
   14:  75%,  75%,  75%
   15:  87%,  85%,  60%
    0:  96%,  96%,  95%
```

I do not quite understand where the color for index 2 is coming from, but the color it represents seems needlessly close to black at index 1. (In fact, I'm pretty sure they are both black in a 4-bits per channel color depth).

Or, am I misunderstanding something and the fact that it is not precisely black a sign that ImageMagick is doing some — uh — _magic_ during the `-remap` operator where it is detecting a duplicate color and deciding to fill it in with something more useful? 

Also, is the fact that we're attempting these various kludges mean that ImageMagick's quantization algorithm simply cannot be primed with a few specific colors? 

_____
On a tangent, I didn't use `-unique-colors` because I wanted to see duplicates, yet I somehow got the correct number of colors, as you can see above. Weird, huh? On the other hand, IM legacy's `-colors` seems to be buggy when writing PNG files:

```bash
$ convert logo: -colors 16 foo.png && identify foo.png
foo.png PNG 640x480 640x480+0+0 8-bit sRGB 17c 17858B 0.000u 0:00.000
$ convert logo: -colors 16 png8:bar.png && identify bar.png
bar.png PNG 640x480 640x480+0+0 8-bit sRGB 16c 17842B 0.000u 0:00.000
```

**Comment by snibgo** [comment]
> but ImageMagick is ignoring the three prechosen colors during quantization and could easily choose a very similar color.

Correct.

> I do not quite understand where the color for index 2 is coming from, but the color it represents seems needlessly close to black at index 1. 

Colour index 2 (near black) was one of the "-colors 13". The 13 did not have a pure black. As a general rule, "-colors" will not choose pure (aka "extreme") colors. Besides, "logo:" has no black pixels, so there is no chance that "-colors 13" will include a black.

> it is detecting a duplicate color and deciding to fill it in with something more useful?

No. IM won't do that.

> ImageMagick's quantization algorithm simply cannot be primed with a few specific colors?

Correct.

Also note that "-colors 13" may choose less than 13 colours. This can even occur when 13 could be used. If we want exactly 13 colours, workarounds are possible, but messy.

Also note that "-colors" and "-remap" work in sRGB space, which is perceptually non-uniform. Instead we can do that in a more uniform colorspace such as CIELab for a visually better result.

**Comment by snibgo** [comment]
> On a tangent, I didn't use -unique-colors because I wanted to see duplicates, yet I somehow got the correct number of colors, as you can see above. Weird, huh?

"-remap" effectively does a "-unique-colors" first. You can use operations like "+write x.png" to capture intermediate results.

You still haven't said what version of IM you are using. If v7, I suggest using "magick" instead of "convert".

```
magick logo: -colors 16 x.png

magick x.png info:

x.png PNG 640x480 640x480+0+0 8-bit sRGB 17c 14192B 0.000u 0:00.002
```
"17c" in this case doesn't mean the image has 17 colours.

```
magick x.png -verbose info:

:
:
  Colors: 16
:
:
  Colormap entries: 17
:
:
```
The colormap (aka palette) has 17 entries, but only 16 are used.

**Reply by hackerb9** [reply]
I happen to be trying with ImageMagick 6 on Unix, but I'll use whatever works. 

Because there are so few colors and because dithering is not a good idea with sixels, choosing the best colors is important for this application. 

How does one use CIELab colorspace for quantization? Anthony's documentation seemed to say that `-colorspace Lab` wouldn't work correctly for the `-posterize` operation you had recommended. Colorspace is one of the reasons I had been using `-colors` instead.  

Is CIELab recommended over YUV? I had been planning on using YUV since that's what is suggested as looking "empirically better" in https://legacy.imagemagick.org/script/quantize.php. (Perhaps that documentation needs to be updated? Also, it would be nice if at the beginning where it starts talking about "RGB space" if there was a note about the `-colorspace` option.)

**Comment by snibgo** [comment]
Sadly, many IM operations only work when the channels represent R,G and B. If this isn't true, IM converts the image to sRGB first. So the trick is to convert the image to Lab (or LUV or whatever you want) with "-colorspace Lab", but then use "-set colorspace sRGB" which will change metadata only. Then IM processes think the image is already sRGB so won't convert it. At the end, before writing the output, reverse the process with "-set colorspace Lab -colorspace sRGB".

In my experience, for ordinary photos CIELab gives perceptually better results than sRGB. For graphics which have highly saturated colours, the opposite is often true. Luv gives a similar appearance to Lab, and I have no preference between these.

**Reply by hackerb9** [reply]
Sounds like a bug in ImageMagick if it is converting to RGB for quantization when it is not necessary (nor desirable). Has this already been reported?
